File: src/JavaScript/Sorting/insertion_sort.js
class InsertionSort {

    /**
     * Uses insertion sort to sort an array into ascending order.
     */
    sortAscending(array) {
        const length = array.length;
        if (length === 0)
            return [];
        for (let j = 1; j < length; j++) {
            const key = array[j];
            let i = j - 1;
            while (i >= 0 && array[i] > key) {
                array[i + 1] = array[i];
                i--;
            }
            array[i + 1] = key;
        }
        return array;
    }

    /**
     * Uses insertion sort to sort an array into descending order.
     */
    sortDescending(array) {
        const length = array.length;
        if (length === 0)
            return [];
        for (let j = 1; j < length; j++) {
            const key = array[j];
            let i = j - 1;
            while (i >= 0 && array[i] < key) {
                array[i + 1] = array[i];
                i--;
            }
            array[i + 1] = key;
        }
        return array;
    }
}

module.exports = InsertionSort;

File: src/Python/Sorting/insertion_sort.py
class InsertionSort:
    
    def sort_ascending(self, array):
        length = len(array)
        if (length == 0):
            return []
        for j in range(1, length):
            key = array[j]
            i = j - 1
            while (i >= 0 and array[i] > key):
                array[i + 1] = array[i]
                i -= 1
            array[i + 1] = key
        return array
    
    def sort_descending(self, array):
        length = len(array)
        if (length == 0):
            return []
        for j in range(1, length):
            key = array[j]
            i = j - 1
            while (i >= 0 and array[i] < key):
                array[i + 1] = array[i]
                i -= 1
            array[i + 1] = key
        return array